Initialized app and implemented get emp data

anastaklo opened the pull request:
Implements https://github.com/anastaklo/employee-portal/issues/1

aklochkova committed:
created pom

anastaklo committed:
added modules structure
added Person domain object

anastaklo committed:
cleared pom.xml
rearranged modules

anastaklo committed:
refactored modules
added empty classes for Person use case

anastaklo committed:
reshaped modules structure

anastaklo committed:
fixed path to parent pom

anastaklo committed:
renamed Person to Employee
added JPA entities
added connection to db
created init script for db
implemented 'get' for employee data

anastaklo committed:
made modelMapper private

marpozh commented on .gitignore:
```diff
@@ -21,3 +21,6 @@
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+
+#IntellijIdea
```
/.idea/
/target/
*.log
*.iml

anastaklo replied:
`*.log` is on the 5 line and `*.iml` on the 26, I will add `/.idea/` and `/target/`

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/BaseJpaEntity.java:
```diff
@@ -0,0 +1,34 @@
+    @Override
+    public int hashCode() {
+
+        return 13;
```
what the magic?)

anastaklo replied:
I borrowed this idea from this article https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/, see section `Fixing the entity identifier equals and hashCode`

> there is only one solution: the hashCode should always return the same value

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/CalendarJpaEntity.java:
```diff
@@ -0,0 +1,38 @@
+    @OneToMany(mappedBy = "parentCalendar")
+    @Getter
+    @Setter
+    private List<CalendarJpaEntity> childCalendars = new ArrayList<>();
```
why should it know about children?

anastaklo replied:
honestly I don't know :) we need to discuss calendar specific part (maybe in another task, or I can remove this mapping if it's definitely unnecessary)

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/CalendarJpaEntity.java:
```diff
@@ -0,0 +1,38 @@
+    @Getter
+    @Setter
+    @NonNull
+    private List<EmployeeJpaEntity> employees = new ArrayList<>();
```
not the single employee?

anastaklo replied:
oh, maybe it's a mistake, we really need to discuss calendars theme again...

marpozh replied:
ок)

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/DocumentJpaEntity.java:
```diff
@@ -0,0 +1,33 @@
+    @NonNull
+    @Getter
+    @Setter
+    private EventJpaEntity event;
```
hm, i think a document - is the separate entity and an event should khow about own documents.
also some documents in the future will be not "link" with events.

anastaklo replied:
do you mean why I placed here `@ManyToOne`? I made it for better performance to prevent Hibernate from extra updates (https://vladmihalcea.com/the-best-way-to-map-a-onetomany-association-with-jpa-and-hibernate/)
I agree about linkage to events, so will remove `@NonNull`

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/EmployeeJpaEntity.java:
```diff
@@ -0,0 +1,120 @@
+@Table(name = "employee")
+public class EmployeeJpaEntity extends BaseJpaEntity{
+
+    @Column
```
constraints?
like @Size?

anastaklo committed:
Merge remote-tracking branch 'origin/master' into init-application

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/EmployeeJpaEntity.java:
```diff
@@ -0,0 +1,120 @@
+    private String firstName;
+
+    @Column
+    @Getter
```
@Getter
@Setter for class-level

marpozh replied:
@EqualsAndHashCode for all classes?

anastaklo replied:
no, we can't use `@EqualsAndHashCode` for the vast majority of our entities because I chose automatically generated by database `id`, so this `id` differs in different entity states (it will be `null` when entity is in `transient` state and later it will be set to some number after persisting)

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/PositionJpaEntity.java:
```diff
@@ -0,0 +1,34 @@
+    @Getter
+    @Setter
+    @NonNull
+    private String category; //todo: maybe change this field to just 'isManager'?
```
i think it's doesn't matter for us. but 'category' is easier to support if we add other categories in the future

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/RoomJpaEntity.java:
```diff
@@ -0,0 +1,39 @@
+    @Column
+    @Getter
+    @Setter
+    private String floor;
```
string?

anastaklo replied:
I thought about something like '12A' but maybe it's an overthinking :) will change it to `int`

marpozh commented on adapter/persistence/src/main/java/org/anastaklo/portal/adapter/persistence/VacationRequestJpaEntity.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public int hashCode() {
+
+        return 13;
```
why not 42?)

anastaklo replied:
please see my previous answer

anastaklo replied:
I will document it for further readability

marpozh commented on configuration/src/main/resources/db/migration/V20200416__init-schema.sql:
```diff
@@ -0,0 +1,134 @@
+CREATE TABLE IF NOT EXISTS Position
```
not 'position' ?

anastaklo replied:
do you mean we need to lowercase all table names? no problem

marpozh commented on configuration/src/main/resources/db/migration/V20200416__init-schema.sql:
```diff
@@ -0,0 +1,134 @@
+CREATE TABLE IF NOT EXISTS Position
+(
```
I feel we need indexes - don't forget in the future

anastaklo replied:
I'll create a separate task for it

anastaklo committed:
placed gitbook in documentation folder

marpozh commented on domain/src/main/java/org/anastaklo/portal/entities/Employee.java:
```diff
@@ -0,0 +1,19 @@
+
+@Data
+public class Employee {
+    
```
maybe id wil be helpfull for update operation

anastaklo committed:
removed diagrams from gitbook

anastaklo committed:
added GlobalExceptionHandler.java
renamed db tables
added 'id' to Employee.java
added folders to .gitignore
added constraints to jpa entities

anastaklo committed:
Merge remote-tracking branch 'origin/master' into init-application

anastaklo committed:
moved gitbook files

anastaklo committed:
added scopes to methods

anastaklo committed:
replace @Autowired properties with constructors
